From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: William Blake Galbreath <Blake.Galbreath@GMail.com>
Date: Fri, 3 Jul 2020 11:58:56 -0500
Subject: [PATCH] Add PrepareResultEvent / PrepareGrindstoneEvent

Adds a new event for all crafting stations that generate a result slot item

Anvil, Grindstone and Smithing now extend this event

Grindstone is a backwards compat from a previous PrepareGrindstoneEvent

diff --git a/src/main/java/io/papermc/paper/event/inventory/PrepareGrindstoneEvent.java b/src/main/java/io/papermc/paper/event/inventory/PrepareGrindstoneEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..70244efa97b81c8b7f08c932b9d05ad4173b4186
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/inventory/PrepareGrindstoneEvent.java
@@ -0,0 +1,28 @@
+package io.papermc.paper.event.inventory;
+
+import org.bukkit.Warning;
+import org.bukkit.inventory.GrindstoneInventory;
+import org.bukkit.inventory.InventoryView;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Called when an item is put in a slot for grinding in a Grindstone
+ * @deprecated use {@link org.bukkit.event.inventory.PrepareGrindstoneEvent}
+ */
+@Deprecated
+@Warning(false)
+public class PrepareGrindstoneEvent extends PrepareResultEvent {
+
+    public PrepareGrindstoneEvent(@NotNull InventoryView inventory, @Nullable ItemStack result) {
+        super(inventory, result);
+    }
+
+    @NotNull
+    @Override
+    public GrindstoneInventory getInventory() {
+        return (GrindstoneInventory) super.getInventory();
+    }
+
+}
diff --git a/src/main/java/io/papermc/paper/event/inventory/PrepareResultEvent.java b/src/main/java/io/papermc/paper/event/inventory/PrepareResultEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..36d9dd97c13867f7f7a7c141c69db36989b4cc4f
--- /dev/null
+++ b/src/main/java/io/papermc/paper/event/inventory/PrepareResultEvent.java
@@ -0,0 +1,37 @@
+package io.papermc.paper.event.inventory;
+
+import org.bukkit.event.inventory.PrepareInventoryResultEvent;
+import org.bukkit.inventory.InventoryView;
+import org.bukkit.inventory.ItemStack;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Called when an item is put in an inventory containing a result slot
+ */
+public class PrepareResultEvent extends PrepareInventoryResultEvent {
+
+    // HandlerList on PrepareInventoryResultEvent to ensure api compat
+    public PrepareResultEvent(@NotNull InventoryView inventory, @Nullable ItemStack result) {
+        super(inventory, result);
+    }
+
+    /**
+     * Get result item, may be null.
+     *
+     * @return result item
+     */
+    @Nullable
+    public ItemStack getResult() {
+        return super.getResult();
+    }
+
+    /**
+     * Set result item, may be null.
+     *
+     * @param result result item
+     */
+    public void setResult(@Nullable ItemStack result) {
+        super.setResult(result);
+    }
+}
diff --git a/src/main/java/org/bukkit/event/inventory/PrepareAnvilEvent.java b/src/main/java/org/bukkit/event/inventory/PrepareAnvilEvent.java
index 6782024735a885ba0b1b4dba4a576740c1410366..087c38b97161c7d28c923ffa607b3eeb23255553 100644
--- a/src/main/java/org/bukkit/event/inventory/PrepareAnvilEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/PrepareAnvilEvent.java
@@ -10,9 +10,9 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Called when an item is put in a slot for repair by an anvil.
  */
-public class PrepareAnvilEvent extends PrepareInventoryResultEvent {
+public class PrepareAnvilEvent extends io.papermc.paper.event.inventory.PrepareResultEvent {
 
-    private static final HandlerList handlers = new HandlerList();
+    // Paper - move HandlerList to PrepareInventoryResultEvent
 
     public PrepareAnvilEvent(@NotNull InventoryView inventory, @Nullable ItemStack result) {
         super(inventory, result);
@@ -24,14 +24,5 @@ public class PrepareAnvilEvent extends PrepareInventoryResultEvent {
         return (AnvilInventory) super.getInventory();
     }
 
-    @NotNull
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    @NotNull
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
+    // Paper - move HandlerList to PrepareInventoryResultEvent
 }
diff --git a/src/main/java/org/bukkit/event/inventory/PrepareGrindstoneEvent.java b/src/main/java/org/bukkit/event/inventory/PrepareGrindstoneEvent.java
index fb172479ce28edbf969b9492236e30fb04395bf9..749be86687033a8968d0ac026356aaa7ac36a6e0 100644
--- a/src/main/java/org/bukkit/event/inventory/PrepareGrindstoneEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/PrepareGrindstoneEvent.java
@@ -10,9 +10,9 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Called when an item is put in a slot for repair or unenchanting in a grindstone.
  */
-public class PrepareGrindstoneEvent extends PrepareInventoryResultEvent {
+public class PrepareGrindstoneEvent extends io.papermc.paper.event.inventory.PrepareGrindstoneEvent { // Paper
 
-    private static final HandlerList handlers = new HandlerList();
+    // Paper - move HandlerList to PrepareInventoryResultEvent
 
     public PrepareGrindstoneEvent(@NotNull InventoryView inventory, @Nullable ItemStack result) {
         super(inventory, result);
@@ -24,14 +24,5 @@ public class PrepareGrindstoneEvent extends PrepareInventoryResultEvent {
         return (GrindstoneInventory) super.getInventory();
     }
 
-    @NotNull
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    @NotNull
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
+    // Paper - move HandlerList to PrepareInventoryResultEvent
 }
diff --git a/src/main/java/org/bukkit/event/inventory/PrepareInventoryResultEvent.java b/src/main/java/org/bukkit/event/inventory/PrepareInventoryResultEvent.java
index b543bc17fb9354d1939c67c1fb6c92d88fdbe4ec..7368857c34fcd1ef62c4dfd07d5ffc71175a9d72 100644
--- a/src/main/java/org/bukkit/event/inventory/PrepareInventoryResultEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/PrepareInventoryResultEvent.java
@@ -8,7 +8,9 @@ import org.jetbrains.annotations.Nullable;
 
 /**
  * Called when an item is put in a slot and the result is calculated.
+ * @deprecated use {@link io.papermc.paper.event.inventory.PrepareResultEvent}
  */
+@Deprecated @org.bukkit.Warning(false) // Paper
 public class PrepareInventoryResultEvent extends InventoryEvent {
 
     private static final HandlerList handlers = new HandlerList();
diff --git a/src/main/java/org/bukkit/event/inventory/PrepareSmithingEvent.java b/src/main/java/org/bukkit/event/inventory/PrepareSmithingEvent.java
index 901774e03f8789dddff4e7695ac599ff69cc97a5..99a25103cf1496d10ee1ca9fa8de81bd1b006702 100644
--- a/src/main/java/org/bukkit/event/inventory/PrepareSmithingEvent.java
+++ b/src/main/java/org/bukkit/event/inventory/PrepareSmithingEvent.java
@@ -10,9 +10,9 @@ import org.jetbrains.annotations.Nullable;
 /**
  * Called when an item is put in a slot for upgrade by a Smithing Table.
  */
-public class PrepareSmithingEvent extends PrepareInventoryResultEvent {
+public class PrepareSmithingEvent extends io.papermc.paper.event.inventory.PrepareResultEvent {
 
-    private static final HandlerList handlers = new HandlerList();
+    // Paper - move HandlerList ot PrepareInventoryResultEvent
 
     public PrepareSmithingEvent(@NotNull InventoryView inventory, @Nullable ItemStack result) {
         super(inventory, result);
@@ -24,14 +24,5 @@ public class PrepareSmithingEvent extends PrepareInventoryResultEvent {
         return (SmithingInventory) super.getInventory();
     }
 
-    @NotNull
-    @Override
-    public HandlerList getHandlers() {
-        return handlers;
-    }
-
-    @NotNull
-    public static HandlerList getHandlerList() {
-        return handlers;
-    }
+    // Paper - move HandlerList to PrepareInventoryResultEvent
 }
